lfs/setup.go

```diff
@@ -1,160 +1,80 @@
-		"#!/bin/sh\ngit lfs push --stdin \"$@\"": true,
-		"#!/bin/sh\ngit lfs pre-push \"$@\"":     true,
-		"#!/bin/sh\ncommand -v git-lfs >/dev/null 2>&1 || { echo >&2 \"\\nThis repository has been set up with Git LFS but Git LFS is not installed.\\n\"; exit 0; }\ngit lfs pre-push \"$@\"": true,
-		"#!/bin/sh\ncommand -v git-lfs >/dev/null 2>&1 || { echo >&2 \"\\nThis repository has been set up with Git LFS but Git LFS is not installed.\\n\"; exit 2; }\ngit lfs pre-push \"$@\"": true,
-	}
-)
-
-type HookExists struct {
-	Name     string
-	Path     string
-	Contents string
-}
-
-func (e *HookExists) Error() string {
-	return fmt.Sprintf("Hook already exists: %s\n\n%s\n", e.Name, e.Contents)
-}
-
-func InstallHooks(force bool) error {
-	if !InRepo() {
-		return NotInARepositoryError
+	// prePushHook invokes `git lfs push` at the pre-push phase.
+	prePushHook = &Hook{
+		Type:     PrePushHook,
+		Contents: "#!/bin/sh\ncommand -v git-lfs >/dev/null 2>&1 || { echo >&2 \"\\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/pre-push.\\n\"; exit 2; }\ngit lfs pre-push \"$@\"",
+		Upgradeables: []string{
+			"#!/bin/sh\ngit lfs push --stdin $*",
+			"#!/bin/sh\ngit lfs push --stdin \"$@\"",
+			"#!/bin/sh\ngit lfs pre-push \"$@\"",
+			"#!/bin/sh\ncommand -v git-lfs >/dev/null 2>&1 || { echo >&2 \"\\nThis repository has been set up with Git LFS but Git LFS is not installed.\\n\"; exit 0; }\ngit lfs pre-push \"$@\"",
+			"#!/bin/sh\ncommand -v git-lfs >/dev/null 2>&1 || { echo >&2 \"\\nThis repository has been set up with Git LFS but Git LFS is not installed.\\n\"; exit 2; }\ngit lfs pre-push \"$@\"",
+		},
 	}
 
-	if err := os.MkdirAll(filepath.Join(LocalGitDir, "hooks"), 0755); err != nil {
-		return err
+	// hooks is a collection of all hooks to be installed by Git LFS.
+	hooks = []*Hook{
+		prePushHook,
 	}
 
-	hookPath := filepath.Join(LocalGitDir, "hooks", "pre-push")
-	if _, err := os.Stat(hookPath); err == nil && !force {
-		return upgradeHookOrError(hookPath, "pre-push", prePushHook, prePushUpgrades)
-	}
+	// cleanFilter invokes `git lfs clean` as the clean filter.
+	cleanFilter = &Filter{Name: "clean", Value: "git lfs clean %%f"}
 
-	return ioutil.WriteFile(hookPath, []byte(prePushHook+"\n"), 0755)
-}
+	// smudgeFilter invokes `git lfs smudge` as the smudge filter.
+	smudgeFilter = &Filter{Name: "smudge", Value: "git lfs smudge %%f"}
 
-func UninstallHooks() error {
-	if !InRepo() {
-		return NotInARepositoryError
-	}
+	// XXX(@ttaylorr) not sure if this makes sense as a filter? Perhaps a
+	// Settable or Attribute type may be more appropriate.
 
-	prePushHookPath := filepath.Join(LocalGitDir, "hooks", "pre-push")
-	file, err := os.Open(prePushHookPath)
-	if err != nil {
-		// hook doesn't exist, our work here is done
-		return nil
-	}
+	requireFilters = &Filter{Name: "required", Value: "true"}
```

Comment:
I wonder if we're thinking about filters incorrectly? Instead of thinking of it as smudge/clean/required filters, think of it as the `lfs` filter, with values for smudge/clean/required. Check out the docs:

> If a filter must succeed in order to make the stored contents usable, you can declare that the filter is required, in the configuration:
>
>```
[filter "crypt"]
	clean = openssl enc ...
	smudge = openssl enc -d ...
	required
>```

http://git-scm.com/docs/gitattributes